### PR TITLE
Don't overwrite existing files when DOWNLOAD_ALWAYS_ASK is false

### LINKED ISSUE #7764: Downloading the same name file disabling the save file dialog should change the filename
## Test plan

https://github.com/brave/browser-laptop/pull/9867#issue-240628027

----

**Describe the issue you encountered:**
Downloading the same name file disabling the save file dialog should change the filename, like Edge does. For now the file is overwritten, which users do not always want to see.

- Platform (Win7, 8, 10? macOS? Linux distro?): 

- Brave Version (revision SHA): 0.14.0 Preview

- Steps to reproduce:
    1. Open about:preferences
    1. Specify a download path
    2. Disable "Always ask me where to save files"
    1. Download a file of a release from https://github.com/brave/browser-laptop/releases (eg RELEASES-ia32 of the Preview3)
    2. Wait until the download ends
    3. Download another file which has the same name of a **different version**  (eg RELEASES-ia32 of the **Preview2**)
    
- Actual result: the file is overwritten

- Expected result: the file should be saved with the filename changed, eg RELEASES-ia32(1)

- Any related issues: #2110

Marking this with impact/high as the users could lose their files by accident because of this setting, which behaves slightly different than the other browsers.

### PR DESCRIPTION
**Pull request by timborden**
Fixes #7764

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Test Plan:
- Tested that the file was downloaded to the location specified in the settings
- Tested that a second downloaded file would not replace an existing file of the same name

Reviewer Checklist:

Tests

- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

### CODE REVIEW

File: app/filtering.js [hunk 1]
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
     }
 
     const defaultPath = path.join(getSetting(settings.DOWNLOAD_DEFAULT_PATH) || getSetting(settings.DEFAULT_DOWNLOAD_SAVE_PATH) || app.getPath('downloads'), itemFilename)
-    const savePath = ((process.env.SPECTRON || (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation())) ? defaultPath : dialog.showSaveDialog(win, { defaultPath }))
+    let savePath = ((process.env.SPECTRON || (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation())) ? defaultPath : dialog.showSaveDialog(win, { defaultPath }))
```
Comment (bridiver) [inline app/filtering.js]: I think there should be a single `if` statement here that wraps all of this, but can you switch `!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation()` to `getSetting(settings.DOWNLOAD_ALWAYS_ASK)  || item.promptForSaveLocation()`? The former is confusing

File: app/filtering.js [hunk 2]
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
...
 
     // User cancelled out of save dialog prompt
     if (!savePath) {
       event.preventDefault()
       return
     }
 
+    if (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation()) {
```
Comment (bridiver) [inline app/filtering.js]: please merge into `else` for `if` above

File: app/filtering.js [hunk 3]
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
...
+      let willNotOverwrite = false
```
Comment (bridiver) [inline app/filtering.js]: double negative checks are also hard to understand `!willNotOverwrite` -> `willOverwrite`

File: app/filtering.js [hunk 4]
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
...
+      let matchedFilenames = 0
+      let querySavePath
+      while (!willNotOverwrite) {
+        querySavePath = (matchedFilenames ? savePath.substring(0, savePath.lastIndexOf('.') >= 0 ? savePath.lastIndexOf('.') : savePath.length) + ` (${matchedFilenames})` + savePath.substring(savePath.lastIndexOf('.') >= 0 ? savePath.lastIndexOf('.') : savePath.length) : savePath)
```
Comment (bridiver) [inline app/filtering.js]: I think this will break things like `.tar.gz`. Maybe get the first index?

**Comment by timborden** [comment]
Thanks for the feedback @bridiver 

I've refactored the code based on your suggestions.

I also dug into the extension name edge cases (like `tar.gz`). As a result I switched from the `substring()` approach to using `path.extname()`. Sadly, it didn't solve the issue, but the code is shorter and more readable.

I did some searching for possible solutions. One thought was to use the MIME:
```
> var mime = require('mime-types')
undefined
> mime.lookup('package.json')
'application/json'
> mime.extension(mime.lookup('package.json'))
'json'
```
....sadly, it didn't work for `tar.gz` files
```
> mime.lookup('v0.17.13dev.tar.gz')
false
```

It seems that the most common approach is to split on the last "."...even the OS X save prompt defaults to using the last ".":
screen shot 2017-07-06 at 11 01 42

Short of rolling with a whitelist of file extensions, which feels like overkill....it looks like this may be the best option.

**Comment by timborden** [comment]
Just added tests....FYI

**Comment by luixxiul** [comment]
@timborden would you please rebase on the latest master? thanks!

**Comment by timborden** [comment]
Hey @luixxiul ...sorry for the relay (just changed my GitHub settings so I don't miss these notifications)

I just rebased....FYI

**Comment by luixxiul** [comment]
@timborden thanks! I'd like to see you keep up the nice work :-)

**Comment by bsclifton** [comment]
@bridiver did you want to re-review before this gets merged?

**Comment by bsclifton** [comment]
Merged 😄 👌

**Comment by timborden** [comment]
Awesome....thanks @bsclifton, @bridiver and @luixxiul

**Pull request merged**